parser/src/grammar.lalrpop

```diff
@@ -0,0 +1,424 @@
+        ":" => Token::Colon,
+        "," => Token::Comma,
+        "." => Token::Dot,
+        "=" => Token::Equals,
+        "\"" => Token::Lambda,
+        "|" => Token::Pipe,
+        "->" => Token::RightArrow,
+
+        "{" => Token::Open(Delimiter::Brace),
+        "[" => Token::Open(Delimiter::Bracket),
+        "(" => Token::Open(Delimiter::Paren),
+        "}" => Token::Close(Delimiter::Brace),
+        "]" => Token::Close(Delimiter::Bracket),
+        ")" => Token::Close(Delimiter::Paren),
+
+        "OpenBlock" => Token::OpenBlock,
+        "CloseBlock" => Token::CloseBlock,
+        "Semi" => Token::Semi,
+    }
+}
+
+// Utils
+
+Comma<Rule>: Vec<Rule> =
+    <rules: (<Rule> ",")*> <last: Rule?> => {
+        let mut rules = rules;
+        rules.extend(last);
+        rules
+    };
+
+S<Rule>: Spanned<Rule, BytePos> =
+    <l: @L> <rule: Rule> <r: @R> =>
+        pos::spanned2(l, r, rule);
+
+IdentStr: &'input str = {
+    "Identifier" => <>,
+    "(" <"Operator"> ")" => <>,
+};
+pub Ident: Id = {
+    IdentStr => env.from_str(<>)
+};
+
+// Types
+
+TypeBinding: TypeBinding<Id> = {
+    <id: Ident> <params: Ident*> "=" <vs: Variant+> => {
+        let arg_types = params.iter()
+                        .map(|id| {
+                            Type::generic(Generic {
+                                kind: Kind::variable(0),
+                                id: id.clone(),
+                            })
+                        })
+            .collect();
+        let return_type: ArcType<Id> = if params.is_empty() {
+            Type::ident(id.clone())
+        } else {
+            Type::app(Type::ident(id.clone()), arg_types)
+        };
+        let body = Type::variants(vs.into_iter()
+            .map(|(id, tys)| (id, Type::function(tys, return_type.clone())))
+            .collect());
+        TypeBinding {
+            comment: None,
+            name: id.clone(),
+            alias: Alias::new(id,
+                                params.into_iter()
+                                    .map(|id| {
+                                        Generic {
+                                            kind: Kind::variable(0),
+                                            id: id,
+                                        }
+                                    })
+                                    .collect(),
+                                body),
+        }
+    },
+
+    <id: Ident> <params: Ident*> "=" <body: Type> => {
+        TypeBinding {
+            comment: None,
+            name: id.clone(),
+            alias: Alias::new(id,
+                                params.into_iter()
+                                    .map(|id| {
+                                        Generic {
+                                            kind: Kind::variable(0),
+                                            id: id,
+                                        }
+                                    })
+                                    .collect(),
```

Comment:
If you split this into an args variable, it would look much happier. Eg:

```rust
    <id: Ident> <params: Ident*> "=" <body: Type> => {
        let args = params.into_iter()
                         .map(|id| Generic { kind: Kind::variable(0), id: id })
                         .collect(),

        TypeBinding {
            comment: None,
            name: id.clone(),
            alias: Alias::new(id, args, body),
        }
    }
```